src/reporter/subscription_builder.py

```diff
@@ -22,7 +26,7 @@ def build_subscription(quantumleap_url,
         ),
         notification(
             build_notification_target(quantumleap_url, time_index_attribute),
-            metadata(['dateCreated', 'dateModified', 'TimeInstant']),
+            metadata(metadata_list),
```

Comment:
I believe you don't need to change this, right?. All you want to change is in time_index's (`timex.py`) side of QuantumLeap. Afaik, Orion will not interpret any new data in this `metadata` list other than `dateCreated` and `dateModified` (I still need to check what about `TimeInstant`). In other words, this change seems to have no effect to me, then we don't need to update the expected 'metadata' in the tests neither.

Reply:
Orion only sends those metadata attributes that are in list with the notification. See [http://telefonicaid.github.io/fiware-orion/api/v2/stable/](http://telefonicaid.github.io/fiware-orion/api/v2/stable/) -> Specification -> Filtering out attributes and metadata

So if you use <code>'metadata': ['dateCreated', 'dateModified']</code> when making the subscription, <code>timex.py</code> cannot access <code>TimeInstant</code>, <code>timestamp</code> or a custom meta attribute because they are not included in the notification. In that case, it would end up using <code>dateModified</code> but that might be totally different than the other time-related meta attributes.

But when you use <code>'metadata': ['dateCreated', 'dateModified', 'TimeInstant', 'timestamp']</code> when making the subscription, <code>TimeInstant</code> and <code>timestamp</code> will be included in the notification and can be used in determining the time index. Of course, <code>TimeInstant</code> and <code>timestamp</code> will only be included by Orion if they actually exist within the entity. Any custom meta attribute would have to included similarly in the subscription.

Reply:
Cool, I didn't manage to test if `TimeInstant` and `timestamp` had special treatment by Orion. So, in that case, then it's fine to keep them there. I just tested the inclusion of the custom attribute in metadata list of the subscription and noticed that didn't make Orion include it at all. What QL will have to do is to include the custom attribute instead of in the metadata list, in the notification's attrs list. The user should do that of course, but QL's doing it is a safer play since the user already specified it in the header. Does it make sense?

Reply:
AFAIK the <code>dateCreated</code> and <code>dateModified</code> are the only attributes that have special treatment in Orion. I.e. they are automatically created as entity level attributes as well as attribute level meta attributes.

To make use of the step 1 of the selection strategy (i.e. custom attribute), the custom attribute must be included in the notication's attrs list as you said. The same is with the <code>TimeInstant</code> attribute (step 3), the <code>timestamp</code> attribute (step 5) and the <code>dateModified</code> attribute (step 7). Unless they are included in the attrs list, the notification will not include those attributes and obviously those steps in the selection strategy cannot be used. This is similarly to the steps concerning the meta attributes (steps 2, 4, 6 and 8) that need the attribute names in notification metadata list.

Note that empty list (or missing list) means that all attributes (or all metadata) are included in the notification with the exception of the built-in attributes <code>dateCreated</code> and <code>dateModified</code>. To include those they need to be specifically included by the lists as is explained in the link in my previous comment.

Instead of including all those attributes in the notification attrs list, I would rather leave it up to the user. That way the user can have more control on the time index selection. By default i.e. with parameters <code>attributes=None</code> and <code>notified_attributes=None</code> for the subscription builder, the attribute list is omitted, meaning that all attributes will be included in the notification. That of course includes the user selected custom attribute as well.

Reply:
A cause of confusion was the dilemma of whether including an attr in the metadata list of the subscription would add such attribute in the notification or not. We conclude it will add it in the notification but in the metadata section only (meta attribute) provided it existed, not as entity-level attributes (this is just managed by attrs list of the subscription). IIRC, the case of `dateModified` / `dateCreated` is just slightly different because even with an empty list in the attrs list of the subscription (i.e., asking it all), they still are needed in the metadata list if someone wants them to appear as "entity level attributes".

You're right, with default values for the subscription ensuring the custom attribute sent via header will arrive in the notification at entity-level (and metadata level) we're good.

Thanks again for your PR and patience in the discussion :)